cmd/bosun/conf/system.go

```diff
@@ -0,0 +1,406 @@
+// a map of headers to be sent with each Graphite request
+type GraphiteConf struct {
+	Host    string
+	Headers map[string]string
+}
+
+// AnnotateConf contains the elastic configuration to enable Annotations support
+type AnnotateConf struct {
+	Hosts []string // CSV of Elastic Hosts, currently the only backend in annotate
+	Index string   // name of index / table
+}
+
+// LogstashConf contains a list of elastic hosts for the depcrecated logstash functions
+type LogStashConf struct {
+	Hosts expr.LogstashElasticHosts
+}
+
+// ElasticConf contains configuration for an elastic host that Bosun can query
+type ElasticConf struct {
+	Hosts expr.ElasticHosts
+}
+
+// ElasticConf contains configuration for an influx host that Bosun can query
+type InfluxConf struct {
+	URL       URL
+	Username  string
+	Password  string `json:"-"`
+	UserAgent string
+	Timeout   Duration
+	UnsafeSSL bool
+	Precision string
+}
+
+// DBConf stores the connection information for Bosun's internal storage
+type DBConf struct {
+	RedisHost     string
+	RedisDb       int
+	RedisPassword string
+
+	LedisDir      string
+	LedisBindAddr string
+}
+
+// SMTPConf contains information for the mail server for which bosun will
+// send emails through
+type SMTPConf struct {
+	EmailFrom string
+	Host      string
+	Username  string
+	Password  string `json:"-"`
+}
+
+// GetSystemConfProvider returns the SystemConfProvider interface
+// and validates the logic of the configuration. If the configuration
+// is not valid an error is returned
+func (sc *SystemConf) GetSystemConfProvider() (SystemConfProvider, error) {
+	var provider SystemConfProvider = sc
+	if err := ValidateSystemConf(sc); err != nil {
+		return provider, err
+	}
+	return provider, nil
+}
+
+// NewSystemConf retruns a system conf with default values set
+func newSystemConf() *SystemConf {
+	return &SystemConf{
+		CheckFrequency:  Duration{Duration: time.Minute * 5},
+		DefaultRunEvery: 1,
+		HTTPListen:      ":8070",
+		DBConf: DBConf{
+			LedisDir:      "ledis_data",
+			LedisBindAddr: "127.0.0.1:9565",
+		},
+		MinGroupSize: 5,
+		PingDuration: Duration{Duration: time.Hour * 24},
+		OpenTSDBConf: OpenTSDBConf{
+			ResponseLimit: 1 << 20, // 1MB
+			Version:       opentsdb.Version2_1,
+		},
+		SearchSince:      opentsdb.Day * 3,
+		UnknownThreshold: 5,
+	}
+}
+
+// LoadSystemConfigFile loads the system configuration in TOML format. It will
+// error if there are values in the config that were not parsed
+func LoadSystemConfigFile(fileName string) (*SystemConf, error) {
+	sc := newSystemConf()
+	decodeMeta, err := toml.DecodeFile(fileName, &sc)
+	if err != nil {
+		return sc, err
+	}
+	if len(decodeMeta.Undecoded()) > 0 {
+		return sc, fmt.Errorf("undecoded fields in system configuration: %v", decodeMeta.Undecoded())
+	}
+	sc.md = decodeMeta
+	return sc, nil
+}
+
+// LoadSystemConfig is like LoadSystemConfigFile but loads the config from a string
+func LoadSystemConfig(conf string) (*SystemConf, error) {
```

Comment:
Can we reduce duplication here? Maybe LoadFile uses ioutil.ReadFile then calls this method?